File: cpp project/source/BVH.cpp
#include "stdafx.h"

BVH::BVH(Geometry * geometry, int max_leaf_items_p_)
{
	max_leaf_items_ = max_leaf_items_p_;
	leafs_ = nodes_ = max_depth_ = 0;

	int items_count_ = geometry->number_of_faces();
	items_ = geometry->GetTriangles();

	printf("BVH - Build tree\n");
	printf("Start...\n");

	clock_t start, finish;
	start = clock();
	root_ = BuildTree(0, items_count_ - 1, 0, 2); // zaklad zasobnikova metoda
	finish = clock();
	printf("It takes %d seconds\n", (finish - start) / CLOCKS_PER_SEC);

	nodes_++; // jeden si pripocist za root samotny

	// vypis
	printf("Leafs : %d\n", leafs_);
	printf("Nodes : %d\n", nodes_);
	printf("Items : %d\n", items_count_);
	printf("Max d.: %d\n", max_depth_);
	printf("Leaf items: %d\n", max_leaf_items_);
	printf("Bounds=(%.3f, %.3f, %.3f) x (%.3f, %.3f, %.3f)\n",
		 root_->bounding.bounds[0].x, root_->bounding.bounds[0].y, root_->bounding.bounds[0].z,
		 root_->bounding.bounds[1].x, root_->bounding.bounds[1].y, root_->bounding.bounds[1].z);
}

Node * BVH::BuildTree(int from, int to, const int depth, char axis)
{
	const int n = to - from + 1;
	Node * node = new Node(from, to);

	for (int i = from; i <= to; i++)
	{
		node->bounding.Merge(items_[i].Bounds());
	}

	if (n <= max_leaf_items_)
	{
		leafs_++;
		max_depth_ = MAX(depth, max_depth_);
		return node;
	}

	// pokud to neni leaf -> node
	nodes_ += 2;

	const int pivot = n / 2 + from;

	Sort(&items_[from], n, axis);

	// rozdeleni na dva podintervaly
	axis = (axis + 1) % 3;

	// leafy
	node->children[0] = BuildTree(from, pivot - 1, depth + 1, axis);
	node->children[1] = BuildTree(pivot, to, depth + 1, axis);

	return node; // vraci koren
}

void BVH::Traverse(Ray & ray)
{
	Traverse(ray, root_, FLT_MIN, FLT_MAX);
}

void BVH::Traverse(Ray & ray, Node * node, float t0, float t1)
{
	if (RayBoxIntersection(ray, node->bounding, t0, t1))
	{
		if (node->IsLeaf())
		{
			// projdeme trojuhelniky
			for (int i = node->span[0]; i <= node->span[1]; i++) RayTriangleIntersection97(&items_[i], ray);
		}
		else
		{
			// projdeme oba potomky
			Traverse(ray, node->children[0], t0, t1);
			Traverse(ray, node->children[1], t0, t1);
		}
	}
}

File: static/js/factory/bvh.js
raytracer.factory("BVH", [
	"Node",
	"Intersection",
function(
	Node,
	Intersection
) {
	
	var BVH = function(geometry, maxLeafItems) {
		// koren stromu
		this._root = null;

		// vsechny trojuhelniky nactene v geometry
		this._items = null;

		// pocty itemu a leafu
		this._leafs = 0;
		this._nodes = 0;
		this._maxDepth = 0;

		// max. pocet
		this._maxLeafItems = maxLeafItems; // max_leaf_items_

		this._itemsCount = geometry.numberOfFaces();
		this._items = geometry.getTriangles();

		console.log("BVH - Build tree\n");
		console.log("Start...\n");

		// zaklad zasobnikova metoda
		this._root = this.buildTree(0, this._itemsCount - 1, 0, 2);

		// jeden si pripocist za root samotny
		this._nodes++;

		this._debugInfo();
	};

	BVH.prototype.buildTree = function(from, to, depth, axis) {
		var n = to - from + 1;

		var node = new Node(from, to);

		for (var i = from; i <= to; i++)
		{
			node.getBounding().merge(this._items[i].bounds());
		}

		if (n <= this._maxLeafItems)
		{
			this._leafs++;
			this._maxDepth = Math.max(depth, this._maxDepth);

			return node;
		}

		// pokud to neni leaf -> node
		this._nodes += 2;

		var pivot = n / 2 + from;

		Intersection.sort(this._items[from], n, axis); 

		// rozdeleni na dva podintervaly
		axis = (axis + 1) % 3;

		// leafy
		node.addChildren(this.buildTree(from, pivot - 1, depth + 1, axis));
		node.addChildren(this.buildTree(pivot, to, depth + 1, axis));

		return node; // vraci koren
	};

	BVH.prototype.traverse = function(ray) {
		this._traverse(ray, this._root, -Infinity, Infinity);
	};

	BVH.prototype._traverse = function(ray, node, t0, t1) {
		var rbi = Intersection.rayBoxIntersection(ray, node.rwData().bounding, t0, t1);
		if (rbi.test)
		{
			if (node.isLeaf())
			{
				var forFrom = node.rwData().span[0];
				var forTo = node.rwData().span[1];

				// projdeme trojuhelniky
				for (var i = forFrom; i <= forTo; i++) {
					Intersection.rayTriangleIntersection97(this._items[i], ray);
				}
			}
			else
			{
				// projdeme oba potomky
				this._traverse(ray, node.rwData().children[0], rbi.t0, rbi.t1);
				this._traverse(ray, node.rwData().children[1], rbi.t0, rbi.t1);
			}
		}
	};

	BVH.prototype._debugInfo = function() {
		console.log("Leafs : %d\n", this._leafs);
		console.log("Nodes : %d\n", this._nodes);
		console.log("Items : %d\n", this._itemsCount);
		console.log("Max d.: %d\n", this._maxDepth);
		console.log("Leaf items: %d\n", this._maxLeafItems);
		console.log("Bounds=(%.3f, %.3f, %.3f) x (%.3f, %.3f, %.3f)\n",
			this._root.bounding.bounds[0].x,
			this._root.bounding.bounds[0].y,
			this._root.bounding.bounds[0].z,
			this._root.bounding.bounds[1].x,
			this._root.bounding.bounds[1].y,
			this._root.bounding.bounds[1].z
		);
	};

	return BVH;
}]);
